src/test/java/frc/robot/subsystems/ball/intake/IntakeImplTest.java

```diff
@@ -0,0 +1,52 @@
+package frc.robot.subsystems.ball.intake;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import org.mockito.Mockito;
+import org.junit.Before;
+import frc.robot.subsystems.ball.intake.IntakeImpl;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj.SpeedController;
+
+public class IntakeImplTest {
+    SpeedController intakeMotorController;
+    DoubleSolenoid intakeSolenoid;
+    IntakeImpl intakeSubsystem;
+
+    @Before
+    public void setup() {
+        intakeMotorController = mock(SpeedController.class);
+        intakeSolenoid = mock(DoubleSolenoid.class);
+        intakeSubsystem = new IntakeImpl(intakeSolenoid, intakeMotorController);
+    }
+
+    @Test
+    public void testConstructor(){
+        // Constructor is run earlier
+
+        Mockito.when(intakeSolenoid.get()).thenReturn(DoubleSolenoid.Value.kReverse);
+
+        // Check the initial state of the solenoid
+        assertEquals(false, intakeSubsystem.getDeployed());
+    }
+
+    @Test
+    public void testExtend() {
+        // Extend the intake
+        intakeSubsystem.extend();
+
+        // Check if it behaves as expected
+        assertEquals(true, intakeSubsystem.getDeployed());
+    }
+
+    @Test
+    public void testRetract() {
+        // Retract the intake
+        intakeSubsystem.retract();
+
+        // Check behaviour
+        assertEquals(false, intakeSubsystem.getDeployed());
+    }
```

Comment:
You can also use `Mockito.verify()` to check that calling a method on the object to test results in it calling some other methods on mocks. ex. `intakeSubsystem.retract(); intakeSubsystem.task(); Mockito.verify(intakeSolenoid).set(whateverDirection)`.

Reply:
You can also use this same technique to test more complex behaviours - you could for example retract, run task, verify solenoid action, then extend, run task, and verify the second solenoid action. Not a particularly interesting example for this simple subsystem, but still a useful tool.

Reply:
I added some more to the tests using verify(). Right now it doesn't make sense to do any more with this but I'll definitely end up writing more unit tests soon, so now that I know how to do this I can do some more complex behaviour, and structure the tests differently as well.